peps/pep-0797.rst

```diff
@@ -0,0 +1,672 @@
+reference counting, there is no risk of data races. In fact, both
+subinterpreters and free-threading already use immortality for thread-safety
+on some common objects, such as ``None``, ``True``, or ``False``.
+
+Lastly, immortality is also faster than deferred reference counting
+(:c:func:`PyUnstable_Object_EnableDeferredRefcount`), which is used for
+mitigating reference count contention on some objects on the free-threaded build.
+It would be nice to allow users (likely large C API wrappers) to take advantage
+of that speedup.
+
+Deallocators Rely On the Interpreter State
+------------------------------------------
+
+Another, less intuitive, danger of reference counting is that when the
+object's reference count reaches zero, it immediately tries to get freed.
+In doing so, an object will release its memory back to CPython's object allocator
+(typically via :c:func:`PyObject_GC_Del` or :c:func:`PyObject_Free`). This
+allocator is per-interpreter, so it is unsafe to try to release an object's
+memory in an interpreter different from the one that created it.
+
+Immortality Has No Impact on Reference Counting Performance
+-----------------------------------------------------------
+
+If the primary issue with cross-interpreter reference counting is thread-safety,
+why not make reference counting atomic? Unfortunately, atomic operations are
+slower than their non-atomic counterparts, so this would hurt overall
+performance. :pep:`703` managed to solve this issue using biased reference
+counting, splitting the reference count field into two, but at the cost of
+breaking the stable C API. Atomic reference counting on its own also doesn't
+solve the per-interpreter deallocator issue as mentioned previously.
+
+Immortality is a Powerful Tool to Wield
+***************************************
+
+In general, immortal objects seem to have a lot of applications that aren't
+immediately clear to CPython maintainers. For example, `Nanobind used immortal objects`_
+in its original :term:`free threading` implementation. So, it seems like immortality
+has interesting use-cases in third-party libraries.
+
+.. _Nanobind used immortal objects: https://github.com/wjakob/nanobind/pull/695
+
+Specification
+=============
+
+Changes to the Internal API
+***************************
+
+Internally, there are two functions for making an object immortal:
+``_Py_SetImmortal`` and ``_Py_SetImmortalUntracked``.
+Because the interpreter might want to do its own things with immortal
+objects, either now or in the future, the interpreter state should yield
+control of a user-defined immortal object to the internal API. So,
+the existing internal functions will gain an extra case to remove an object
+from the user-defined immortals, so the interpreter won't try and manually
+free it later. A new internal function, ``_Py_SetImmortalKnown``, will be
+used instead for places where it's certain that objects could not have
+been made immortal by a user.
```

Comment:
I feel like this is too much prose for the specification, the discussion over the current internal functions etc. should probably go into *Rationale*. Instead, I would have this section say "We introduce a third function for immortalising an object, :c:func`!_Py_SetImmortalKnown`. This will be used in places where ...".